solvers/Cov-tWDA.py

```diff
@@ -0,0 +1,722 @@
+        ----------
+        covtest : ndarray, shape (n_trials, n_channels, n_channels)
+            ndarray of SPD matrices.
+        Returns
+        -------
+        preds : ndarray of int, shape (n_trials, 1)
+            the prediction for each trials according to the closest centroid.
+        """
+        discrimination = self._predict_discimination(covtest)
+
+        preds = []
+        n_trials, n_classes = discrimination.shape
+        for i in range(n_trials):
+            preds.append(self.classes_[discrimination[i, :].argmax()])
+        preds = np.asarray(preds)
+        return preds
+
+    def transform(self, S):
+        """get the distance to each centroid.
+        Parameters
+        ----------
+        S : ndarray, shape (n_trials, n_channels, n_channels)
+            ndarray of SPD matrices.
+        Returns
+        -------
+        dist : ndarray, shape (n_trials, n_classes)
+            the distance to each centroid according to the metric.
+        """
+        return self._predict_discrimination(S)
+
+    def fit_predict(self, S, y):
+        """Fit and predict in one function."""
+        self.fit(S, y)
+        return self.predict(S)
+
+    def predict_proba(self, S):
+        """Predict proba using softmax.
+        Parameters
+        ----------
+        S : ndarray, shape (n_trials, n_channels, n_channels)
+            ndarray of SPD matrices.
+        Returns
+        -------
+        prob : ndarray, shape (n_trials, n_classes)
+            the softmax probabilities for each class.
+        """
+        return softmax(self._predict_discrimination(S))
+
+#################################
+# Utils for SPD manifold        #
+#################################
+
+
+def sym(x):
+    """
+    Computes the symmetric part of a given matrix
+
+    Parameters
+    ----------
+    x : array
+        Square Matrix.
+
+    Returns
+    -------
+    array
+        Symmetric part of x.
+
+    """
+    return 0.5 * (x + x.T)
+
+
+def generate_random_SPD(p, cond, random_state=None):
+    """
+    Generate pxp SPD matrix of the form U^TDU and with conditionning number
+    where U is pxp matrix drawn uniformly from the orthogonal group,
+    and D is pxp diagonal matrix
+
+    Parameters
+    ----------
+    p : int
+        dimension of the random matrix.
+    cond : float
+        Conditionning number (in the sense of L2 norm)
+    random_state : int, RandomState instance or None
+        Controls the pseudo random number generation
+
+    Returns
+    -------
+    array
+        Random SPD matrix.
+
+    """
+    U = ortho_group.rvs(p, random_state=random_state)
+    d = np.zeros(p)
+    if p > 2:
+        d[:p-2] = uniform.rvs(loc=1/np.sqrt(cond),
+                              scale=np.sqrt(cond)-1/np.sqrt(cond),
+                              size=p-2, random_state=random_state)
+    d[p-2] = 1/np.sqrt(cond)
+    d[p-1] = np.sqrt(cond)
+    return U @ np.diag(d) @ U.T
+
+
+class SPD(RiemannianSubmanifold):

```

Comment:
The issue here is that when `pymanopt` is not installed, this object cannot be instanciated.
This is a common issue with `benchopt`, not super easy to solve, sorry.

I think the best would be to first move these definitions in a separate `benchmark_utils/twda.py` file and import them in the `safe_import_context`.